### PR TITLE
fix(codex): honor explicit image_generation tool choice

### LINKED ISSUE #3018: Codex image_generation tool injection may affect non-Codex harnesses
### Summary

After PR #2962, the Codex executor appears to inject the built-in `image_generation` tool into every Codex upstream request body. This is useful for Codex CLI image generation, but it may have unintended side effects for other OpenAI-compatible harnesses/clients that route through the Codex executor and do not expect this tool to be present.

Could you please confirm whether this unconditional injection is intended for all Codex executor traffic, or whether it should be gated by client/harness/config?

### Regression range

I checked the request body produced by the Codex executor with a fake upstream server:

- `v6.9.34`: does **not** add `image_generation`
- `v6.9.35`: **does** add `image_generation`
- `v6.9.36`: **does** add `image_generation`

The relevant change appears to be:

- PR: #2962
- Commit: `31934ae0 feat(codex): enable image generation for all Codex upstream requests`

### What changed

`ensureImageGenerationTool()` is called from the Codex executor request paths, including streaming requests. When the incoming request already has tools, the executor appends:

```json
{"type":"image_generation","output_format":"png"}
```

For example, a request that originally contains only:

```json
{"tools":[{"type":"web_search_preview"}]}
```

is sent upstream as something like:

```json
{
  "tools": [
    {"type":"web_search"},
    {"type":"image_generation","output_format":"png"}
  ]
}
```

### Why this may be a problem

The PR comments already mention that non-Codex harnesses can expose/list this injected tool and may fail the turn if the model chooses it. I can reproduce the request-body change locally without needing real upstream credentials.

This can present externally as empty responses, stream errors, or failed request accounting even when many normal text requests still work.

### Minimal reproduction

A minimal test can be done by pointing `CodexExecutor.ExecuteStream` at an `httptest` server and capturing the request body sent to upstream.

Observed result:

- On `v6.9.34`, captured upstream body has no `image_generation` tool.
- On `v6.9.35` and `v6.9.36`, captured upstream body includes `image_generation` even when the original request did not include it.

Test command used locally:

```bash
EXPECT_IMAGE_TOOL=0 go test -run TestCodexImageGenerationToolInjectionRepro ./internal/runtime/executor  # passes on v6.9.34
EXPECT_IMAGE_TOOL=1 go test -run TestCodexImageGenerationToolInjectionRepro ./internal/runtime/executor  # passes on v6.9.35/v6.9.36
```

### Suggested fix

Could this be made configurable or gated more narrowly? For example:

- Add a config option to disable Codex `image_generation` tool injection.
- Only inject for known Codex CLI / ChatGPT-compatible clients that actually support the tool.
- Preserve the original behavior by default for generic OpenAI-compatible clients/harnesses.

Thanks!

### LINKED ISSUE #3023: image generation does not work
PS C:\Users\user\Desktop\CLIProxyAPIPlus> curl.exe -v -X POST "http://localhost:8317/v1/images/generations" `
>>   -H "Authorization: Bearer sk-izi-Z2xHQvSC0fdmHstP2UrfQhdlJ-EUoR84njViEJTXFunVhsSf3" `
>>   -H "Content-Type: application/json" `
>>   --data-binary "@img.json"
Note: Unnecessary use of -X or --request, POST is already inferred.
* Host localhost:8317 was resolved.
* IPv6: ::1
* IPv4: 127.0.0.1
*   Trying [::1]:8317...
*   Trying 127.0.0.1:8317...
* Established connection to localhost (::1 port 8317) from ::1 port 59458
* using HTTP/1.x
> POST /v1/images/generations HTTP/1.1
> Host: localhost:8317
> User-Agent: curl/8.16.0
> Accept: */*
> Authorization: Bearer sk-izi-Z2xHQvSC0fdmHstP2UrfQhdlJ-EUoR84njViEJTXFunVhsSf3
> Content-Type: application/json
> Content-Length: 159
>
* upload completely sent off: 159 bytes
< HTTP/1.1 400 Bad Request
< Access-Control-Allow-Headers: *
< Access-Control-Allow-Methods: GET, POST, PUT, PATCH, DELETE, OPTIONS
< Access-Control-Allow-Origin: *
< Content-Type: application/json
< Date: Fri, 24 Apr 2026 16:00:00 GMT
< Content-Length: 183
<
{
  "error": {
    "message": "Tool choice 'image_generation' not found in 'tools' parameter.",
    "type": "invalid_request_error",
    "param": "tool_choice",
    "code": null
  }
}* Connection #0 to host localhost:8317 left intact
PS C:\Users\user\Desktop\CLIProxyAPIPlus>

### PR DESCRIPTION
**Pull request by gaozuo**
Fixes #3018.
Fixes #3023.
Related to #3036.
Related to #2991.

Separate from usage retry accounting in #3042.

## Summary
- stops unconditional Codex `image_generation` tool injection for generic Responses requests
- ensures `tools` contains `image_generation` only when the request explicitly selects the image generation tool
- preserves existing client-provided `image_generation` tool definitions without duplicating or overwriting them
- keeps existing safeguards for `spark` models and free Codex auth
- routes native OpenAI-compatible `/v1/images/generations` requests through providers that explicitly support the native Images endpoint
- prevents custom/native image models from falling back to Codex `gpt-5.4-mini` when no native Images provider is available
- keeps `stream:true` native image requests from being silently downgraded to non-streaming responses
- applies configured payload rules and the existing OpenAI-compatible User-Agent behavior on native Images requests

## Why
Codex clients can expose the native `image_generation` tool in their own tool registry. The proxy should not infer image generation from headers or generic chat text. It should only repair explicit structured image-tool intent, such as a frontend/client selecting the image generation tool.

This avoids the upstream error shape reported in related issues where `tool_choice` selects `image_generation` but the `tools` array does not contain the matching tool.

The OpenAI-compatible Images API is a separate native endpoint family. Provider image models such as Qwen/native Images API models should be routed by the requested image model to a native Images provider, not translated through the Codex Responses image tool fallback.

## Notes on related issues
- #3018 is the core Codex regression: `image_generation` was injected into generic Responses traffic.
- #3023 is covered by adding the missing `image_generation` tool when the request explicitly selects it.
- #3036 is related because it reports the same missing-tool error, but free-account image generation limits are not changed here.
- #2991 is related because this PR fixes the native Images routing side where custom/provider image models could end up requesting Codex `gpt-5.4-mini`; provider/account support limitations are not changed here.

## Local validation
- `go test -count=1 ./internal/runtime/executor -run 'TestOpenAICompatExecutorNativeImages|TestOpenAICompatExecutorSupportsNativeImages'`
- `go test -count=1 ./sdk/api/handlers/openai -run 'TestImages(Generations|Edits)'`
- `go test -count=1 ./internal/runtime/executor ./sdk/api/handlers ./sdk/api/handlers/openai ./sdk/cliproxy/auth ./sdk/cliproxy/executor`
- `go test -race -count=1 ./internal/runtime/executor -run 'TestOpenAICompatExecutorNativeImages|TestOpenAICompatExecutorSupportsNativeImages'`
- `go test -race -count=1 ./sdk/api/handlers/openai -run 'TestImages(Generations|Edits)'`
- `go build -o test-output ./cmd/server && rm test-output`

I also rebuilt a downstream local deploy stack with this patched kernel and verified:
- native image model requests are routed as the requested provider/model instead of `codex/gpt-5.4-mini`
- `gpt-image-2` still works through the Codex image tool fallback and returns a real `b64_json` image payload

### CODE REVIEW

**Comment by gaozuo** [comment]
Updated this PR to address the review feedback:

- Removed the test data race by passing the captured upstream request body through buffered channels instead of sharing a mutable variable or calling `t.Fatalf` inside the HTTP handler goroutine.
- Honored explicit `tool_choice: {"type":"allowed_tools","tools":[{"type":"image_generation"}]}` by injecting the missing image_generation tool.
- Rebasing onto the latest `dev` keeps the PR current.

Validation run locally:

- `go test ./internal/runtime/executor -run 'Test(EnsureImageGenerationTool|MaybeEnsureImageGenerationTool|CodexExecutor.*ImageGeneration)'`
- `go test -race ./internal/runtime/executor -run 'TestCodexExecutor.*ImageGeneration|TestMaybeEnsureImageGenerationTool_AddsForAllowedToolsImageGenerationToolChoice'`
- `go build -o test-output ./cmd/server && rm test-output`

**Review by luispater — CHANGES_REQUESTED** [review]
### Summary
This PR stops unconditional injection of the Codex `image_generation` tool into upstream Responses requests. It now only ensures `tools` contains `image_generation` when the incoming request explicitly selects `tool_choice = image_generation`, while preserving the existing `spark` and free-plan safeguards. Good coverage is added via new/expanded executor tests.

### Blocking
- `internal/runtime/executor/codex_executor_tool_injection_test.go` captures the upstream request body via a shared `capturedBody` variable written in the HTTP handler goroutine and read later without explicit synchronization, and the handler calls `t.Fatalf`. This can be racy and can lead to confusing failures (especially under `go test -race`).
  - Suggested fix: send the captured body (and any read error) back to the test goroutine via a buffered channel (or use a mutex + waitgroup), and only call `t.Fatal/Fatalf` from the main test goroutine.

### Non-blocking
- Add a small unit test for the `"tool_choice": "image_generation"` string form, since `shouldEnsureImageGenerationTool` explicitly supports it.
- (Optional) `ensureImageGenerationTool` now duplicates the scan logic implemented by `hasImageGenerationTool`; consider reusing it to keep one source of truth.

### Test plan
- Not run locally (review-only).
- PR description lists targeted `go test` commands for `./internal/runtime/executor`.

**Comment by gaozuo** [comment]
Updated again to address the maintainer review points directly:

- Blocking data-race concern: `executeCodexStreamAndCaptureBody` now sends the captured upstream request body and any read error back to the main test goroutine through buffered channels. The HTTP handler no longer shares a mutable `capturedBody` variable and no longer calls `t.Fatal/Fatalf` from the handler goroutine.
- String-form coverage: added `TestMaybeEnsureImageGenerationTool_AddsForStringImageGenerationToolChoice` for `"tool_choice": "image_generation"`, matching the code path already supported by `shouldEnsureImageGenerationTool`.
- Single source of truth for tool scanning: added `jsonToolsIncludeImageGeneration` and reused it from `toolChoiceIncludesImageGeneration`, `hasImageGenerationTool`, and `ensureImageGenerationTool` so image tool detection is no longer duplicated.

Validation run locally after the update:

- `go test ./internal/runtime/executor -run 'Test(EnsureImageGenerationTool|MaybeEnsureImageGenerationTool|CodexExecutor.*ImageGeneration)'`
- `go test -race ./internal/runtime/executor -run 'TestCodexExecutor.*ImageGeneration|TestMaybeEnsureImageGenerationTool_(AddsForStringImageGenerationToolChoice|AddsForAllowedToolsImageGenerationToolChoice)'`
- `go build -o test-output ./cmd/server && rm test-output`

**Comment by gaozuo** [comment]
Final update after local end-to-end validation with the patched kernel mounted into the downstream deploy stack.

What changed in this update:

- Kept the maintainer-requested data-race fix in `executeCodexStreamAndCaptureBody`: the HTTP handler now sends captured request bodies/read errors through buffered channels, and all `t.Fatal/Fatalf` calls happen in the test goroutine.
- Kept the `allowed_tools` fix for explicit image generation: `tool_choice: {"type":"allowed_tools","tools":[{"type":"image_generation"}]}` now ensures the top-level `tools` array contains the `image_generation` tool.
- Refactored image-tool detection into one helper (`toolsIncludeImageGeneration`) so `ensureImageGenerationTool`, `hasImageGenerationTool`, and explicit `tool_choice.tools` checks share one source of truth.
- Removed support/test coverage for `"tool_choice": "image_generation"` after real upstream validation showed it is not a valid Codex/Responses shape. The upstream rejects it with: `Invalid value: 'image_generation'. Supported values are: 'none', 'auto', and 'required'.`

Local automated validation:

- `go test ./internal/runtime/executor -run 'Test(EnsureImageGenerationTool|MaybeEnsureImageGenerationTool|CodexExecutor.*ImageGeneration)'`
- `go test -race ./internal/runtime/executor -run 'TestCodexExecutor.*ImageGeneration|TestMaybeEnsureImageGenerationTool_AddsForAllowedToolsImageGenerationToolChoice'`
- `go build -o test-output ./cmd/server && rm test-output`

Real request validation:

I copied this patched branch into a downstream deploy stack via a local module replacement, rebuilt the app image, and sent real requests to the running server. The following legal shapes returned HTTP 200 and produced real image results (`image_generation_call` plus generated image data/result):

- `tool_choice: {"type":"image_generation"}` with streaming Responses.
- `tool_choice: {"type":"allowed_tools","tools":[{"type":"image_generation"}]}` with streaming Responses.
- Existing `tools: [{"type":"image_generation"}]` plus `tool_choice: {"type":"image_generation"}`.
- Existing `tools: [{"type":"image_generation"}]` plus `tool_choice: "required"`.
- Non-streaming Responses with `tool_choice: {"type":"image_generation"}`.
- `/v1/images/generations` with `gpt-image-2` returned b64 image data.

Negative/guard validation:

- A generic Responses request with only `web_search_preview` returned HTTP 200 and did not receive an injected `image_generation` tool.
- `"tool_choice": "image_generation"` was intentionally not kept because real upstream validation returns HTTP 400 for that shape.

One transient upstream `server_error` occurred during an image generation attempt with mixed existing tools, but rerunning the same request twice succeeded with generated image results, so it was not a request-shaping or injection issue.

**Comment by gaozuo** [comment]
Hi @luispater, could you please re-review this PR when you have a chance?

I addressed the maintainer feedback and pushed the updated branch:

- Fixed the test race by using buffered channels for captured request bodies/read errors, with all `t.Fatal/Fatalf` calls kept in the test goroutine.
- Added/fixed coverage for the valid `allowed_tools` image-generation shape.
- Refactored image-tool detection to reuse a single helper.
- Removed the `"tool_choice": "image_generation"` string-form path after real upstream validation showed Codex/Responses rejects it with HTTP 400.

I also rebuilt a downstream deploy stack with this patched kernel and verified real image generation succeeds for the valid shapes, while generic `web_search_preview` requests no longer receive an injected `image_generation` tool.

Latest validation details are summarized here: https://github.com/router-for-me/CLIProxyAPI/pull/3019#issuecomment-4319551261

**Comment by gaozuo** [comment]
Sorry for the noisy previous state of this PR.

I found that the branch had accidentally carried unrelated Redis/protocol-multiplexer commits from an older base, which made the PR conflict with the latest `dev` and caused unrelated bot review noise. I cleaned this up by resetting the PR branch onto the current `dev` and reapplying only the Codex image-generation changes.

Current PR state:

- Single commit: `fbde11a0`
- Merge status: clean / no conflicts
- Files changed are now limited to:
  - `internal/runtime/executor/codex_executor.go`
  - `internal/runtime/executor/codex_executor_imagegen_test.go`
  - `internal/runtime/executor/codex_executor_tool_injection_test.go`

Local validation after the cleanup:

- `go test -count=1 ./internal/runtime/executor -run 'Test(EnsureImageGenerationTool|MaybeEnsureImageGenerationTool|CodexExecutor.*ImageGeneration)'`
- `go test -race -count=1 ./internal/runtime/executor -run 'TestCodexExecutor.*ImageGeneration|TestMaybeEnsureImageGenerationTool_AddsForAllowedToolsImageGenerationToolChoice'`
- `go build -o test-output ./cmd/server && rm test-output`

The earlier end-to-end request validation summary still applies to the same image-generation behavior, but the branch is now scoped only to this PR's intended Codex image-tool fix.

**Comment by gaozuo** [comment]
Hi @luispater, the PR branch has now been cleaned up and is ready for re-review/merge when you have time.

Current state:

- Merge conflicts are resolved; GitHub reports the PR as mergeable.
- The branch now contains a single scoped commit: `fbde11a0`.
- The diff is limited to the Codex image-generation tool handling files only:
  - `internal/runtime/executor/codex_executor.go`
  - `internal/runtime/executor/codex_executor_imagegen_test.go`
  - `internal/runtime/executor/codex_executor_tool_injection_test.go`
- All GitHub checks are green.

Local validation after the cleanup:

- `go test -count=1 ./internal/runtime/executor -run 'Test(EnsureImageGenerationTool|MaybeEnsureImageGenerationTool|CodexExecutor.*ImageGeneration)'`
- `go test -race -count=1 ./internal/runtime/executor -run 'TestCodexExecutor.*ImageGeneration|TestMaybeEnsureImageGenerationTool_AddsForAllowedToolsImageGenerationToolChoice'`
- `go build -o test-output ./cmd/server && rm test-output`

Sorry again for the earlier noisy branch state with unrelated commits. This PR should now be image-tool-only and ready for your review.

**Comment by gaozuo** [comment]
Updated this PR with a second, focused fix for native OpenAI-compatible image generation routing.

### Root cause
The `/v1/images/generations` handler previously treated image requests as Codex Responses `image_generation` tool requests by default. That is correct for the built-in Codex image tool path (`gpt-image-2`), but it is wrong for native OpenAI-compatible Images API models such as custom/provider image models. Those requests could be forced through the Codex adapter (`gpt-5.4-mini` + `image_generation`) instead of going to the provider selected by the requested image model.

### What changed
- Added an optional executor capability contract: `executor.NativeImagesEndpointExecutor`.
- `OpenAICompatExecutor` implements that capability for `images/generations` and forwards directly to upstream `/images/generations` without chat/responses translation.
- `/v1/images/generations` now first checks whether the requested image model has a native Images provider.
  - If yes, it routes to that provider's native Images endpoint.
  - If no and the model is not `gpt-image-2`, it returns `502 no native image provider` instead of falling back to Codex.
  - `gpt-image-2` keeps the existing Codex Responses image tool adapter path.
- `/v1/images/edits` still does not claim native edit support; custom edit models without implemented native edit forwarding are rejected instead of being accidentally sent through the Codex fallback.

### Boundary / design note
This intentionally avoids provider-name hardcoding beyond excluding the Codex fallback path from native Images routing. The capability lives in `sdk/cliproxy/executor`, the executor contract layer, so handlers only ask whether an executor supports a native endpoint. No Business/usage-layer packages or external application concepts are imported.

### Test coverage added
- Native image generation routing by requested model.
- `gpt-image-2` prefers a native Images provider when one exists.
- OpenAI-compatible alias resolution preserves the resolved upstream model while forwarding the original payload shape.
- Registered non-native image providers are rejected instead of called.
- Custom image models do not fall back to Codex/gpt-5.4-mini.
- Unknown custom image models return the native-provider error.
- Custom image edit models are rejected when native edit forwarding is not implemented.
- `OpenAICompatExecutor` forwards native image generation to `/images/generations` and overrides the upstream model correctly.
- Image request detail lookup still keeps `gpt-image-2` blocked from normal chat/responses routes but allowed for images routes.

### Validation run locally
- `go test -count=1 ./internal/runtime/executor ./sdk/api/handlers ./sdk/api/handlers/openai ./sdk/cliproxy/auth ./sdk/cliproxy/executor`
- `go test -race -count=1 ./internal/runtime/executor -run 'Test(OpenAICompatExecutorNativeImages|OpenAICompatExecutorSupportsNativeImages|EnsureImageGenerationTool|MaybeEnsureImageGenerationTool|CodexExecutor.*ImageGeneration)'`
- `go test -race -count=1 ./sdk/api/handlers ./sdk/api/handlers/openai ./sdk/cliproxy/auth ./sdk/cliproxy/executor`
- `go build -o test-output ./cmd/server && rm test-output`

I also ran a real local deploy-stack validation with a production-dump-derived local DB:
- Requesting `qwen/qwen3.5-397b-a17b` via `/v1/images/generations` now records/runs as `provider=nvidia model=qwen/qwen3.5-397b-a17b`, not `provider=codex model=gpt-5.4-mini`.
- Requesting `gpt-image-2` still succeeds through the Codex image tool fallback and returns a real `b64_json` image payload.

`go test ./...` was also attempted; it still fails only on an unrelated existing registry assertion: `TestCodexStaticModelsIncludeGPT55/free` expects the Codex free tier to include `gpt-5.5`.

**Comment by gaozuo** [comment]
Follow-up update after the latest automated review comments.

Addressed the two Codex bot concerns on commit `2c299e62`:

- **Preserve stream behavior**: native `/images/generations` routing is now only used for non-streaming requests. If a custom/native image model sends `"stream": true` and native streaming is not implemented, the handler returns a clear `502 no native image streaming provider` instead of silently converting the request into a non-streaming JSON response.
- **Apply payload rules**: the native OpenAI-compatible images path now applies the same configured payload rules (`default`, `override`, `filter`) used by the normal OpenAI-compatible executor path before forwarding upstream.

Also cleaned up the OpenAI-compatible `User-Agent` handling:

- Reused a single `openAICompatUserAgent` constant instead of repeating the literal `cli-proxy-openai-compat` across normal, native images, and streaming paths.
- This keeps the native images path aligned with the existing OpenAI-compatible executor behavior rather than introducing a new UA.

Additional regression tests added:

- User-Agent consistency for native image generation forwarding.
- Payload rules applied to native image generation requests.
- `stream:true` custom/native image models do not get downgraded to non-streaming native calls.

Validation run locally:

- `go test -count=1 ./internal/runtime/executor -run 'TestOpenAICompatExecutorNativeImages|TestOpenAICompatExecutorSupportsNativeImages'`
- `go test -count=1 ./sdk/api/handlers/openai -run 'TestImages(Generations|Edits)'`
- `go test -count=1 ./internal/runtime/executor ./sdk/api/handlers ./sdk/api/handlers/openai ./sdk/cliproxy/auth ./sdk/cliproxy/executor`
- `go test -race -count=1 ./internal/runtime/executor -run 'TestOpenAICompatExecutorNativeImages|TestOpenAICompatExecutorSupportsNativeImages'`
- `go test -race -count=1 ./sdk/api/handlers/openai -run 'TestImages(Generations|Edits)'`
- `go build -o test-output ./cmd/server && rm test-output`

CodeRabbit local review:

- `coderabbit review --agent -t uncommitted --base upstream/dev -c AGENTS.md` completed with `0` findings before this follow-up commit.
- A post-commit re-run against the same branch was attempted but hit the CodeRabbit account rate limit (`waitTime: 58 minutes and 15 seconds`).

**Comment by gaozuo** [comment]
Hi @luispater, could you please take another look when you have time?

Current state:

- Rebased on the latest `dev`; GitHub reports the PR as mergeable with no conflicts.
- All current GitHub checks are passing.
- The earlier maintainer feedback and the latest automated review comments have been addressed.
- I added regression tests for the Codex image tool behavior and the native Images API routing edge cases (`stream:true`, payload rules, User-Agent consistency, custom model fallback prevention).

Thanks!

**Comment by gaozuo** [comment]
Added one more test-only follow-up commit to make the Codex image tool fix fully covered across the real executor request paths.

New commit: `61d1b341 test(codex): cover image tool injection across request paths`

What changed:

- Expanded the upstream-body capture test from streaming-only coverage into a matrix over all three modified Codex request paths:
  - `Execute` -> `/responses`
  - `ExecuteStream` -> `/responses`
  - `Execute` with `Alt: responses/compact` -> `/responses/compact`
- Each path now captures the actual body sent to the fake upstream and verifies:
  - generic `web_search_preview` clients do not receive `image_generation`
  - plain generic requests do not receive `image_generation`
  - explicit `tool_choice.type = image_generation` receives the missing tool
  - `allowed_tools` with `image_generation` receives the missing tool
  - existing `image_generation` tool definitions are preserved and not duplicated
  - free Codex auth does not inject the image tool
  - spark models do not inject the image tool

Validation rerun locally:

- `go test -count=1 -v ./internal/runtime/executor -run 'TestCodexExecutorImageGenerationToolInjectionAcrossRequestPaths'`
- `go test -count=1 ./internal/runtime/executor -run 'Test(EnsureImageGenerationTool|MaybeEnsureImageGenerationTool|CodexExecutor.*ImageGeneration|CodexExecutorImageGenerationToolInjectionAcrossRequestPaths|OpenAICompatExecutorNativeImages|OpenAICompatExecutorSupportsNativeImages)'`
- `go test -count=1 ./sdk/api/handlers/openai -run 'TestImages(Generations|Edits)'`
- `go test -race -count=1 ./internal/runtime/executor -run 'Test(EnsureImageGenerationTool|MaybeEnsureImageGenerationTool|CodexExecutor.*ImageGeneration|CodexExecutorImageGenerationToolInjectionAcrossRequestPaths|OpenAICompatExecutorNativeImages|OpenAICompatExecutorSupportsNativeImages)'`
- `go test -race -count=1 ./sdk/api/handlers/openai -run 'TestImages(Generations|Edits)'`
- `go test -count=1 ./internal/runtime/executor ./sdk/api/handlers ./sdk/api/handlers/openai ./sdk/cliproxy/auth ./sdk/cliproxy/executor`
- `go build -o test-output ./cmd/server && rm test-output`

**Comment by gaozuo** [comment]
@codex review

File: sdk/api/handlers/openai/openai_images_handlers.go [hunk 1]
```diff
@@ -214,9 +214,37 @@ func (h *OpenAIAPIHandler) ImagesGenerations(c *gin.Context) {
...
+				Message: fmt.Sprintf("no native image provider for model %s", imageModel),
+				Type:    "invalid_request_error",
+			},
+		})
+		return
+	}
 
 	tool := []byte(`{"type":"image_generation","action":"generate"}`)
-	tool, _ = sjson.SetBytes(tool, "model", imageModel)
+	tool, _ = sjson.SetBytes(tool, "model", defaultImagesToolModel)
```
[the remark this answers is not in the document]
Comment (gaozuo) [reply sdk/api/handlers/openai/openai_images_handlers.go]: Thanks for the review. I’m going to keep the canonical fallback here instead of preserving the prefix.

For the Codex Responses image-tool fallback, the public `/v1/images/*` image model may be provider-qualified (for example `openai/gpt-image-2`), but the fallback execution still needs to route through the canonical Codex main model (`gpt-5.4-mini`) and canonical image tool model (`gpt-image-2`). Preserving the prefix here would require a corresponding `prefix/gpt-5.4-mini` route to exist. In real local validation with a restored deploy stack, preserving the `openai/` prefix caused `502 unknown provider for model openai/gpt-5.4-mini`, while canonicalizing to `gpt-5.4-mini` succeeds and generates images.

Native/provider image models are handled separately by the native Images endpoint path and keep their requested provider model there. This branch is only for the built-in Codex image-tool fallback, so dropping the prefix is intentional to avoid routing drift.

File: sdk/api/handlers/openai/openai_images_handlers.go [hunk 2]
```diff
@@ -499,17 +547,7 @@ func (h *OpenAIAPIHandler) imagesEditsFromJSON(c *gin.Context) {
...
-		toolModel := strings.TrimSpace(gjson.GetBytes(toolJSON, "model").String())
-		if idx := strings.LastIndex(toolModel, "/"); idx > 0 && idx < len(toolModel)-1 {
-			prefix := strings.TrimSpace(toolModel[:idx])
-			if prefix != "" {
-				mainModel = prefix + "/" + defaultImagesMainModel
-			}
-		}
-	}
-	req, _ = sjson.SetBytes(req, "model", mainModel)
+	req, _ = sjson.SetBytes(req, "model", defaultImagesMainModel)
```
[the remark this answers is not in the document]
Comment (gaozuo) [reply sdk/api/handlers/openai/openai_images_handlers.go]: Thanks for flagging this. I’m intentionally not restoring the previous prefix-derived main model logic.

That older prefix-preserving logic was not a safe routing contract for this fallback: if a request used `openai/gpt-image-2`, it derived `openai/gpt-5.4-mini`, which can be unregistered and fail before execution. I reproduced exactly that failure in the local deploy stack (`502 unknown provider for model openai/gpt-5.4-mini`). The Codex image-tool fallback should use the canonical main model (`gpt-5.4-mini`) and the canonical tool model (`gpt-image-2`); native OpenAI-compatible image providers are routed through the separate `/images/generations` native path.

So this hardcoding is deliberate: it prevents namespaced image-tool aliases from being misinterpreted as namespaced Codex main-model routes.

**Comment by ZqinKing** [comment]
Please merge them quickly; gpt-image-2 is completely unusable right now.

**Comment by gaozuo** [comment]
老大给俺看下呗，俺第一次提交，有啥问题直接跟俺说俺改 😄 @luispater

**Comment by gaozuo** [comment]
Closing this larger PR for now. I plan to follow up with smaller, focused PRs based on the latest dev branch so each issue is easier to review and merge.
